engineering/architecture/dateandtime.md

```diff
@@ -0,0 +1,107 @@
+---
+title: Human Readable Dates and Times
+description: Spec for handling dates and times in a human friendly way.
+published: true
+date: 2022-02-14T12:35:23.307Z
+tags: 
+editor: markdown
+dateCreated: 2022-02-15T14:58:57.900Z
+---
+
+## Goals
+
+1. Users should be able to interact with dates, times, datetimes, and time durations in Mathesar in the most human-friendly way possible
+2. The UI should be extremely forgiving w.r.t. date/time input, and also display dates and times in a human-friendly way.
+3. The API should be extremely forgiving w.r.t. date/time input, but should return dates, times, and time durations in a more programmatic, i.e., consistent way than the UI.
+4. The UI and API should be consistent in the following way:  If the input string is understood by both UI and API, they should always result in the _same date/time_ being stored, regardless of the input method.
+5. The dates/times understood by the API should be a subset of those understood by the UI. That is, there should be no dates/times/time durations that are understood by the API, but which can't be entered through the UI.
```

Comment:
I don't think the API needs to be restricted based on the UI input.

> understood by the UI

The UI needs only to understand dates & times after the user enters and saves it, in which scenario the backend would provide the same in a well-defined format. See my [comment above](https://github.com/centerofci/mathesar-wiki/pull/35/files#r809144659).

Reply:
I could definitely be convinced to keep the UI layer pretty "thin" for this. That would certainly make keeping things consistent easier.

Reply:
Ah, @pavish I just realized the source of confusion.  By "understood by the UI", I _don't_ mean it needs to be handled by code at the UI layer.  I mean that from a user perspective, the UI should accept all input that the API accepts.  This implies avoiding invalidating some string that the API could have accepted.  I did want to leave the possibility open that the UI could have some fancy parsing that would allow accepting _more_ input styles, and then parse them into RFC 3339 compliant strings to submit to the API.

I'll think about how to word that more clearly.